[CoreIPC] [Fuzz Blocker] TRAP in WebKit::NetworkConnectionToWebProcess::dispatchMessage

pratiksha-99 opened the pull request:
#### d7e44e3524f840f1b08f494db29421c50b2d79d6
<pre>
[CoreIPC] [Fuzz Blocker] TRAP in WebKit::NetworkConnectionToWebProcess::dispatchMessage
<a href="https://bugs.webkit.org/show_bug.cgi?id=282604">https://bugs.webkit.org/show_bug.cgi?id=282604</a>
<a href="https://rdar.apple.com/138840919">rdar://138840919</a>

Reviewed by Chris Dumez.

Added a check for decoder.destinationID() validity on the invalid message received for fuzzing.

* LayoutTests/ipc/coreipc.js: Added.
(deepCopy):
(CoreIPCClass):
(CoreIPCClass.prototype.initializeMessageByName):
(CoreIPCClass.prototype.initializeMessages):
(CoreIPCClass.prototype.generateSendingFunction):
(CoreIPCClass.prototype.newStreamConnection):
(export.StreamConnection):
(export.StreamConnection.prototype.newInterface):
(export.StreamConnectionInterface):
(export.StreamConnectionInterface.prototype.initializeMessages):
(export.StreamConnectionInterface.prototype.generateStreamSendingFunction):
(export.SerializationError):
(export.resolveAlias):
(isPrimtiveType):
(isEnum):
(isIdentifier):
(export.ArgumentSerializer.splitTemplateType):
(export.ArgumentSerializer.parseTemplate):
(export.ArgumentSerializer):
* LayoutTests/ipc/invalid-message-to-web-process-crash-expected.txt: Added.
* LayoutTests/ipc/invalid-message-to-web-process-crash.html: Added.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::dispatchMessage):

Canonical link: <a href="https://commits.webkit.org/286256@main">https://commits.webkit.org/286256@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22dcb4d1af5245ad6c16b0b3770e0be1516a4f20

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75311 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54750 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28151 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/79822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26575 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63891 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2535 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59116 "33 flakes 110 failures") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/17342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78378 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49287 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64704 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39476 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46683 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22197 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/24903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67698 "") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22536 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81258 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2641 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1666 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67365 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2792 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64700 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66641 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16594 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10593 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8760 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2602 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5429 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2627 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3557 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2640 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -288,6 +288,7 @@ bool NetworkConnectionToWebProcess::dispatchMessage(IPC::Connection& connection,
 #endif
 
     if (decoder.messageReceiverName() == Messages::NetworkTransportSession::messageReceiverName()) {
+        MESSAGE_CHECK_WITH_RETURN_VALUE(AtomicObjectIdentifier<WebTransportSessionIdentifierType>::isValidIdentifier(decoder.destinationID()), false);
```
Should just use `WebTransportSessionIdentifier::isValidIdentifier()`

cdumez commented on LayoutTests/ipc/coreipc.js:
```diff
@@ -0,0 +1,1189 @@
+    {type: 'uint64_t', value: 0x41},
+];
+
+class CoreIPCClass {
```
If we want to include this test, it needs to be minimized. This feels like adding an entire fuzzer as a test :)

rniwa replied:
@cdumez : oh, this is the new framework fuzzing team has created. We're expected to keep this file up-to-date with fuzzer team's as an abstraction layer.

cdumez replied:
Ah ok.

cdumez reviewed (approved):
LGTM with the nit fix I suggested then.

cdumez commented on LayoutTests/ipc/invalid-message-to-web-process-crash.html:
```diff
@@ -0,0 +1,19 @@
+    testRunner.waitUntilDone();
+    setTimeout(() => {
+        testRunner.notifyDone();
+    }, 5000);
```
I don't think we should have a test that runs for 5 seconds. That's crazy. Can we make this a lot shorter?

rniwa replied:
10ms should be long enough. Or better yet, delete calls to `testRunner.waitUntilDone` and `testRunner.notifyDone` altogether?

pratiksha-99 committed:
[CoreIPC] [Fuzz Blocker] TRAP in WebKit::NetworkConnectionToWebProcess::dispatchMessage
https://bugs.webkit.org/show_bug.cgi?id=282604
rdar://138840919

Reviewed by Chris Dumez.

Added a check for decoder.destinationID() validity on the invalid message received for fuzzing.

* LayoutTests/ipc/coreipc.js: Added.
(deepCopy):
(CoreIPCClass):
(CoreIPCClass.prototype.initializeMessageByName):
(CoreIPCClass.prototype.initializeMessages):
(CoreIPCClass.prototype.generateSendingFunction):
(CoreIPCClass.prototype.newStreamConnection):
(export.StreamConnection):
(export.StreamConnection.prototype.newInterface):
(export.StreamConnectionInterface):
(export.StreamConnectionInterface.prototype.initializeMessages):
(export.StreamConnectionInterface.prototype.generateStreamSendingFunction):
(export.SerializationError):
(export.resolveAlias):
(isPrimtiveType):
(isEnum):
(isIdentifier):
(export.ArgumentSerializer.splitTemplateType):
(export.ArgumentSerializer.parseTemplate):
(export.ArgumentSerializer):
* LayoutTests/ipc/invalid-message-to-web-process-crash-expected.txt: Added.
* LayoutTests/ipc/invalid-message-to-web-process-crash.html: Added.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::dispatchMessage):

Canonical link: https://commits.webkit.org/286256@main
